extensions/oplogPopulator/OplogPopulator.js

```diff
@@ -27,80 +32,388 @@ class OplogPopulator {
-        config.database = database;
-        // get credentials
-        let cred = '';
-        if (authCredentials &&
-            authCredentials.username &&
-            authCredentials.password) {
-            const username = encodeURIComponent(authCredentials.username);
-            const password = encodeURIComponent(authCredentials.password);
-            cred = `${username}:${password}@`;
-        }
-        let connectionUrl = `mongodb://${cred}${replicaSetHosts}/` +
-            `?w=${writeConcern}&readPreference=${readPreference}`;
-        if (replicaSet) {
-            connectionUrl += `&replicaSet=${replicaSet}`;
-        }
-        config['connection.uri'] = connectionUrl;
+        config.name = connectorName;
+        config.database = this._database;
+        config['connection.uri'] = this._mongoUrl;
         // destination topic configuration
         config['topic.namespace.map'] = JSON.stringify({
             '*': this._config.topic,
         });
         return config;
     }
 
+    /**
+     * Get buckets that have at least one extension active
+     * @returns {string[]} list of buckets to listen to
+     * @throws {InternalError}
+     */
+    async _getBackbeatEnabledBuckets() {
+        const filter = {
+            $or: [],
+        };
+        // notification filter
+        if (this._activeExtensions.includes('notification')) {
+            const field = constants.extensionConfigField.notification;
+            const extFilter = {};
+            // notification config should be an object
+            extFilter[`value.${field}`] = { $type: 3 };
+            filter.$or.push(extFilter);
+        }
+        // replication filter
+        if (this._activeExtensions.includes('replication')) {
+            const field = constants.extensionConfigField.replication;
+            const extFilter = {};
+            // processing buckets where at least one replication
+            // rule is enabled
+            extFilter[`value.${field}.rules`] = {
+                $elemMatch: {
+                    enabled: true,
+                },
+            };
+            filter.$or.push(extFilter);
+        }
+        // lifecycle filter
+        if (this._activeExtensions.includes('lifecycle')) {
+            const field = constants.extensionConfigField.lifecycle;
+            const extFilter = {};
+            // processing buckets where at least one lifecycle
+            // rule is enabled
+            extFilter[`value.${field}.rules`] = {
+                $elemMatch: {
+                    ruleStatus: 'Enabled',
+                },
+            };
+            filter.$or.push(extFilter);
+        }
+        // ingestion filter
+        if (this._activeExtensions.includes('ingestion')) {
+            const field = constants.extensionConfigField.ingestion;
+            const extFilter = {};
+            extFilter[`value.${field}.status`] = 'enabled';
+            filter.$or.push(extFilter);
+        }
```

Comment:
ideally, I would prefer this logic (e.g. the mapping from "extension" to associated filter) to be outside of this code : possibly be part of each extension itself.

So that we don't need to modify oplogPopulator itself, and simply need to add the required "support" in the extension.

Reply:
Techdebt ticket created